WIP to get working with current gradle

bbarker opened the pull request:
- Include's Nanoware's gradle updates
- Adds version of Scala to hello message.
- Still doesn't work: #4

Cervator committed:
Major Gradle overhaul - bunch of stuff done, stable again but more to do.

- Upgrades to Gradle 3.3 from 2.10
- Removes the need for module directories to have a build.gradle
- Related cleanup / simplification of scripts
- Bakes in support for alternative module languages into new /modules Gradle scripts via new "language" property in module.txt

Cervator committed:
Minor cleanup.

bbarker committed:
typo fox; added version to hello for better runtime info

Cervator commented:
I just pushed a few more changes to the Nanoware engine repo. Could you try that setup with this ScalaLib and see if it works?

Reminder to make sure you use a separate workspace with the Nanoware setup since it is vastly different than the regular engine, when it comes to Gradle :-)

I've gone back to pure Gradle for now along with Atom for editing, ignoring IntelliJ until I've worked out some more kinks. I can `gradlew jar game` and do the scala test. I think the modules are also building better in general, although at present binary module dependencies are not being copied into `/modules` so have to use source dirs for everything. Aiming to re-fix that soon.

Cervator commented:
Oh, also, one more thing: I have a `gradle.properties` in the root of my local workspace that may matter. Two parts to it, some config that may otherwise cause trouble (configureondemand will break the current setup), and some stuff that ties it more cleanly to the independent Nanoware test flow (separate Artifactory homes and module management for repos under the Nanoware org instead of the Terasology org)

```
org.gradle.daemon=true
org.gradle.parallel=false
org.gradle.configureondemand=false
alternativeResolutionRepo=http://artifactory.terasology.org/artifactory/virtual-nanoware-and-remote
artifactoryPublishRepo=nanoware-snapshot-local
artifactoryResolveRepo=virtual-nanoware-and-remote
alternativeGithubHome=Nanoware
```

(some of the initial 3 may be defaults, but anyway)

bbarker commented:
I think I tried all these suggestions and still had the same result; the only different I can think of is that I'm using 2.11.11 in both ScalaLib and Terasology repos, but I was getting the exact same error before I started using that, so I doubt that is the issue ... hmm:

```
* CoreSampleGameplay has extra dependencies:
** Core
*** Identified source: Core
Oh hey, a Scala module. Using its build
common.gradle picked up a custom resolution repo: http://artifactory.terasology.org/artifactory/virtual-nanoware-and-remote
Changing PUBLISH repoKey to nanoware-snapshot-local
Changing RESOLVE repoKey to virtual-nanoware-and-remote
Scanning for dependencies in module.txt for ScalaLib
Dependencies: []
Version for ScalaLib loaded as 2.11.11-SNAPSHOT for group org.terasology.modules

Processing module 'ScalaLib' in a multi-project workspace
* No extra dependencies
```
...

```
17:53:12.374 [main] INFO  o.t.e.prefab.internal.PrefabFormat - Attempting to deserialize prefab engine:explosiveDamage with inputs [/engine/assets/prefabs/damageTypes/explosiveDamage.prefab]
17:53:12.580 [main] ERROR o.t.engine.ComponentSystemManager - Failed to load system ScalaLib:ScalaCommands
java.lang.NoClassDefFoundError: scala/collection/Seq
        at java.lang.Class.getDeclaredConstructors0(Native Method)
        at java.lang.Class.privateGetDeclaredConstructors(Class.java:2671)
        at java.lang.Class.getConstructor0(Class.java:3075)
        at java.lang.Class.newInstance(Class.java:412)
        at org.terasology.engine.ComponentSystemManager.loadSystems(ComponentSystemManager.java:96)
        at org.terasology.engine.modes.loadProcesses.RegisterSystems.step(RegisterSystems.java:53)
        at org.terasology.engine.modes.StateLoading.update(StateLoading.java:242)
        at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:413)
        at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:368)
        at org.terasology.engine.Terasology.main(Terasology.java:152)
Caused by: java.lang.ClassNotFoundException: Failed to find or load class scala.collection.Seq
        at org.terasology.module.sandbox.ModuleClassLoader.findClass(ModuleClassLoader.java:158)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at org.terasology.module.sandbox.ModuleClassLoader.loadClass(ModuleClassLoader.java:121)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 10 common frames omitted
Caused by: javassist.NotFoundException: scala.collection.Seq
        at javassist.ClassPool.get(ClassPool.java:452)
        at org.terasology.module.sandbox.ModuleClassLoader$1.run(ModuleClassLoader.java:145)
        at org.terasology.module.sandbox.ModuleClassLoader$1.run(ModuleClassLoader.java:142)
        at java.security.AccessController.doPrivileged(Native Method)
        at org.terasology.module.sandbox.ModuleClassLoader.findClass(ModuleClassLoader.java:142)
        ... 13 common frames omitted
17:53:12.586 [main] INFO  o.t.e.m.l.InitialiseWorld - World seed: "l18rDfTnociVRvprFhqO96y6UWlsqn4F"
17:53:12.818 [main] INFO  o.t.w.b.internal.BlockManagerImpl - Registered BlockFamily[engine:unloaded]
```
FWIW, this is on Linux.

Git status:

```
brandon@brandon-750-170se:~/workspace/Terasology$ git log | head -20
commit 5544f86ab4eec3d3045058890a3d996a3e624502
Author: Cervator <cervator@gmail.com>
Date:   Wed Nov 1 19:53:50 2017 -0400

    Assortment of tweaks to make things work better, committed before I manage to erase them accidentally again. Includes Gradle 4 and some related changes.

commit 56a744f5aa469f1f1f3137bfa595cf50f8610244
Author: Cervator <cervator@gmail.com>
Date:   Tue Oct 31 01:00:06 2017 -0400

    Actually don't depend on the telemetry stuff yet, just make sure Gradle knows where to find it if needed via module->engine snapshot quirk

commit 9499a0df67d140f9031683cddc24e9abc65fcf32
Author: Cervator <cervator@gmail.com>
Date:   Mon Oct 30 23:12:31 2017 -0400

    More remembered things: while the new & faster resolution stuff works in pure Gradle, it makes IntelliJ barf with an unintelligible error on a Gradle refresh. However, oddly, the same exact thing on a testCompile instead of a regular compile source set works fine.

commit 7b0e30ab666cab03a42374995782264a41d0dd1e
Author: Cervator <cervator@gmail.com>
```

```
brandon@brandon-750-170se:~/workspace/Terasology/modules/ScalaLib$ git log | head -20
commit 77c7f771face534cd355f9946fc7ab0d5aa4f90d
Author: bbarker <beb82@cornell.edu>
Date:   Sun Oct 29 21:36:30 2017 -0400

    typo fox; added version to hello for better runtime info

commit 78fb8a0d5b398bceff7b2386d3e536c26d5c4088
Author: Cervator <cervator@gmail.com>
Date:   Fri Jan 27 23:58:56 2017 -0500

    Minor cleanup.

commit 93b6a10ea1c2477656c0032cab13532b41b88f73
Author: Cervator <cervator@gmail.com>
Date:   Thu Jan 26 02:04:36 2017 -0500

    Major Gradle overhaul - bunch of stuff done, stable again but more to do.
    
    - Upgrades to Gradle 3.3 from 2.10
    - Removes the need for module directories to have a build.gradle
```

Only uncomitted change:

```
brandon@brandon-750-170se:~/workspace/Terasology/modules/ScalaLib$ git diff
diff --git a/module.txt b/module.txt
index 558079e..e091d93 100644
--- a/module.txt
+++ b/module.txt
@@ -1,7 +1,7 @@
 {
     "language"          : "scala",
     "id"                : "ScalaLib",
-    "version"           : "2.11.7-SNAPSHOT",
+    "version"           : "2.11.11-SNAPSHOT",
     "author"            : "skaldarnar",
     "displayName"       : "Scala Runtime Library Module",
     "description"       : "Library module that supplies the Scala runtime library. See www.scala-lang.org. Version tied to embedded Scala.",

```

Cervator commented:
Yeah the version won't be it. I actually switched to 2.11.11 as well when you brought it up and it is working fine. The version is set in the `scala.gradle` though, the version entry in the `module.txt` is just for the ScalaLib *module* and I figured it might make sense to keep it in sync with the Scala version used.

Actually, the whole ScalaLib module may become obsolete with the Nanoware changes. Previously it held the *magic* `build.gradle` that added the dependency on Scala, that other modules would then get by depending on ScalaLib. But now just indicating the language in `module.txt` uses the `scala.gradle` file instead - so individual modules would themselves get a direct dependency on Scala.

So in the end you should just need the `moduleExperiment` branch from Nanoware/Terasology, and really all the ScalaLib does is add that one console command for testing it. Can you confirm if you've tried that branch in its own workspace? With the ScalaLib present under `/modules` :-)

bbarker commented:
I switched to the moduleExperiment branch of Terasology (wasn't on it
before), did clean, jar, then game - same problem. I'm not exactly sure
what we mean by a separate workspace; I added Nanoware as a remote, checked
out a branch from Nanoware, and ran gradlew clean; git status reports a
clean tree as well.

On Tue, Nov 7, 2017 at 11:55 PM, Rasmus Praestholm <notifications@github.com
> wrote:

> Yeah the version won't be it. I actually switched to 2.11.11 as well when
> you brought it up and it is working fine. The version is set in the
> scala.gradle though, the version entry in the module.txt is just for the
> ScalaLib *module* and I figured it might make sense to keep it in sync
> with the Scala version used.
>
> Actually, the whole ScalaLib module may become obsolete with the Nanoware
> changes. Previously it held the *magic* build.gradle that added the
> dependency on Scala, that other modules would then get by depending on
> ScalaLib. But now just indicating the language in module.txt uses the
> scala.gradle file instead - so individual modules would themselves get a
> direct dependency on Scala.
>
> So in the end you should just need the moduleExperiment branch from
> Nanoware/Terasology, and really all the ScalaLib does is add that one
> console command for testing it. Can you confirm if you've tried that branch
> in its own workspace? With the ScalaLib present under /modules :-)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Terasology/ScalaLib/pull/5#issuecomment-342710878>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA37jpqndspe4SAuSuUP8ubh8lzUKuKUks5s0TRVgaJpZM4QKkrR>
> .
>



-- 
Brandon Barker
brandon.barker@gmail.com

bbarker commented:
well, more accurately i probably git reset --hard nanoware/master
originally - i seem to recall a problem with checkout.

On Thu, Nov 9, 2017 at 8:02 PM, Brandon Barker <brandon.barker@gmail.com>
wrote:

> I switched to the moduleExperiment branch of Terasology (wasn't on it
> before), did clean, jar, then game - same problem. I'm not exactly sure
> what we mean by a separate workspace; I added Nanoware as a remote, checked
> out a branch from Nanoware, and ran gradlew clean; git status reports a
> clean tree as well.
>
> On Tue, Nov 7, 2017 at 11:55 PM, Rasmus Praestholm <
> notifications@github.com> wrote:
>
>> Yeah the version won't be it. I actually switched to 2.11.11 as well when
>> you brought it up and it is working fine. The version is set in the
>> scala.gradle though, the version entry in the module.txt is just for the
>> ScalaLib *module* and I figured it might make sense to keep it in sync
>> with the Scala version used.
>>
>> Actually, the whole ScalaLib module may become obsolete with the Nanoware
>> changes. Previously it held the *magic* build.gradle that added the
>> dependency on Scala, that other modules would then get by depending on
>> ScalaLib. But now just indicating the language in module.txt uses the
>> scala.gradle file instead - so individual modules would themselves get a
>> direct dependency on Scala.
>>
>> So in the end you should just need the moduleExperiment branch from
>> Nanoware/Terasology, and really all the ScalaLib does is add that one
>> console command for testing it. Can you confirm if you've tried that branch
>> in its own workspace? With the ScalaLib present under /modules :-)
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/Terasology/ScalaLib/pull/5#issuecomment-342710878>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AA37jpqndspe4SAuSuUP8ubh8lzUKuKUks5s0TRVgaJpZM4QKkrR>
>> .
>>
>
>
>
> --
> Brandon Barker
> brandon.barker@gmail.com
>



-- 
Brandon Barker
brandon.barker@gmail.com

Cervator commented:
Different workspace as in an independent clone / copy paste of the first one. Yeah you're likely to hit various fun with Git etc by trying to switch :-)

Hmm, I wonder what it could be. I'm hoping to find @skaldarnar @oniatus or @kaen to eventually look at it and/or the unit testing problem since it *sounds* similar (an unexpected failure of the sandbox) - some more details for that on https://github.com/Nanoware/Terasology/pull/87

Might take a little while but we'll stick with it!

bbarker commented:
I can think of a couple of obvious things to still try:

1. First I just want to double check what commit # you are on for ScalaLib, and run on that if it is different.
2. I can wipe my workspace and start over from a fresh clone
3. Is the fact that `./engine/src/main/java/org/terasology/engine/module/ExternalApiWhitelist.java` doesn't include the scala libs a concern?

Cervator commented:
1 - Good call. I nearly thought that was it for a moment, as I had two local commits in my ScalaLib, until I realized that was just the two commits unique to Nanoware and my origin compared against was Terasology. Oh well. I am on:

```
commit 78fb8a0d5b398bceff7b2386d3e536c26d5c4088
Author: Cervator <cervator@gmail.com>
Date:   Fri Jan 27 23:58:56 2017 -0500

    Minor cleanup.
```

2 - I would just keep a secondary workspace as a good practice when overall workspace structure may vary wildly. That's in part because I'm still a bit scared of Git though ... :D If you're comfortable enough juggling you might be fine in one workspace.

3 - Yes and no. Sometimes new whitelist entries are needed, and this certainly is the sort of thing that usually would involve that. However ... ScalaLib has worked in the past as-is, and whatever witchcraft is making my workspace work now is doing so without a whitelist entry (I double-checked). In some cases so long as a lib doesn't do anything involving core Java stuff (like `File`) it can survive just fine inside module-land.

I do kinda wonder if somehow I've (partially?) broken something about the sandbox setup as part of my changes. Maybe in a way that isn't obvious like a build / project structure quirk that hasn't gone into source control but persists in my workspace.

bbarker commented:
Closing this for now, to continue on Nanoware for now - to summarize on Slack, it was likely an issue with a .gitignored build.gradle sitting in ScalaLib (though I'm not 100% sure).
